### PR TITLE
Dot

### PR DESCRIPTION
**Pull request by pbrunet**

### CODE REVIEW

File: pythran/pythonic/numpy/dot.hpp [hunk 1]
```diff
@@ -1,60 +1,189 @@
...
+    struct is_blas_type : pythonic::types::is_complex<T>
+    {};
+
+template<>
+    struct is_blas_type<float> : std::integral_constant<bool, true>
+    {};
+
+template<>
+    struct is_blas_type<double> : std::integral_constant<bool, true>
+    {};
```
Comment (serge-sans-paille) [inline pythran/pythonic/numpy/dot.hpp]: http://en.cppreference.com/w/cpp/types/is_floating_point

Comment (pbrunet) [inline pythran/pythonic/numpy/dot.hpp]: On 27/03/2015 22:56, serge-sans-paille wrote:
>
> In pythran/pythonic/numpy/dot.hpp 
> <https://github.com/serge-sans-paille/pythran/pull/417#discussion_r27333981>:
>
> >  #include "pythonic/types/traits.hpp"
> >
> > +#include <nt2/linalg/details/blas/mm.hpp>
> > +#include <nt2/linalg/details/blas/mv.hpp>
> > +
> > +template <class T>
> > +    struct is_blas_type : pythonic::types::is_complex<T>
> > +    {};
> > +
> > +template<>
> > +    struct is_blas_type<float> : std::integral_constant<bool, true>
> > +    {};
> > +
> > +template<>
> > +    struct is_blas_type<double> : std::integral_constant<bool, true>
> > +    {};
>
> http://en.cppreference.com/w/cpp/types/is_floating_point
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/serge-sans-paille/pythran/pull/417/files#r27333981>.
>
I am not sure I can use it as it accepts long double while blas do not 
support it.

**Comment by pbrunet** [comment]
This one is ready. It only need a fix from nt2 so that we can use gemv operator.

File: .travis.yml [hunk 2]
```diff
@@ -16,6 +16,7 @@ install:
     - sudo apt-get install -qq libboost-python1.53-dev libgoogle-perftools-dev libgmp-dev libboost1.53-dev git cmake
 
     - sudo update-alternatives --install /usr/bin/gcc gcc /usr/bin/gcc-4.8 80 --slave /usr/bin/g++ g++ /usr/bin/g++-4.8
+    - sudo apt-get install gfortran libopenblas-dev liblapack-dev
```
Comment (serge-sans-paille) [inline .travis.yml]: gfortran! really?

File: pythran/pythonic/numpy/dot.hpp [hunk 3]
```diff
@@ -1,60 +1,189 @@
...
                 {
                     return e*f;
                 }
 
         template<class E, class F>
-            typename std::enable_if<types::is_numexpr_arg<E>::value and types::is_numexpr_arg<F>::value
-                                    and types::numpy_expr_to_ndarray<E>::N == 1 and types::numpy_expr_to_ndarray<F>::N ==1,
-                                    decltype(std::declval<typename types::numpy_expr_to_ndarray<E>::T>()*std::declval<typename types::numpy_expr_to_ndarray<F>::T>())
+            typename std::enable_if<types::is_numexpr_arg<E>::value and types::is_numexpr_arg<F>::value // Arguments are array_like
+                                    and E::value == 1 and F::value ==1, // It is a two vectors.
```
Comment (serge-sans-paille) [inline pythran/pythonic/numpy/dot.hpp]: in the comment, you mean ``dot product of two vectors '' I guess :-)

File: pythran/pythonic/numpy/dot.hpp [hunk 4]
```diff
@@ -1,60 +1,189 @@
...
+                                    typename __combined<typename E::dtype, typename F::dtype>::type
                                    >::type
-                dot(E const& e, F const& f) {
-                    return sum(e*f);
+                dot(E const& e, F const& f)
+                {
+                    return sum(types::numpy_expr<operator_::proxy::mul, E, F>(e, f));
```
Comment (serge-sans-paille) [inline pythran/pythonic/numpy/dot.hpp]: why using operator_ here?

File: pythran/pythonic/numpy/dot.hpp [hunk 5]
```diff
@@ -1,60 +1,189 @@
...
+        template<class E>
+            typename std::enable_if<is_blas_type<E>::value,
+                                    types::ndarray<E, 1>
+                                    >::type
+            dot(types::ndarray<E, 2> const& f, types::ndarray<E, 1> const& e)
+            {
+                types::ndarray<E, 1> out(types::array<long, 1>{{f.shape[0]}}, __builtin__::None);
+                E al = 1, be = 0;
+                const int m = f.shape[1], n = f.shape[0], inc = 1;
+                nt2::details::gemv("T", &m, &n, &al, f.buffer, &m, e.buffer, &inc, &be, out.buffer, &inc);
```
Comment (serge-sans-paille) [inline pythran/pythonic/numpy/dot.hpp]: I guess the "T" means transpose?

File: pythran/pythonic/numpy/dot.hpp [hunk 6]
```diff
@@ -1,60 +1,189 @@
...
+            typename std::enable_if<(not is_blas_type<typename E::dtype>::value or not is_blas_type<typename F::dtype>::value)
+                                    and E::value == 1 and F::value == 2, // And it is vect / matrix
+                                    types::ndarray<typename __combined<typename E::dtype, typename F::dtype>::type, 1>
+                                    >::type
+            dot(E const& e, F const& f)
+            {
+                types::ndarray<typename __combined<typename E::dtype, typename F::dtype>::type, 1> out(types::array<long, 1>{{f.shape[1]}}, 0);
+                for(long i=0; i<out.shape[0]; i++)
+                    for(long j=0; j<f.shape[0]; j++)
+                        out[i] += e[j] * f[types::array<long, 2>{{j, i}}];
```
Comment (serge-sans-paille) [inline pythran/pythonic/numpy/dot.hpp]: you could at least call sum here

File: pythran/toolchain.py [hunk 7]
```diff
@@ -82,8 +83,10 @@ def get(*x):
 
 def _python_ldflags():
     pylibs = sysconfig.get_config_var('LIBS').split()
-    return (["-L" + sysconfig.get_config_var("LIBPL")] +
-            pylibs +
+    numpy_blas = numpy_sys.get_info("blas")
```
Comment (serge-sans-paille) [inline pythran/toolchain.py]: out of curiosity: does this catch the openblas?

File: .travis.yml (hunk 2 above)
Comment (pbrunet) [inline .travis.yml]: If I correctly understand, numpy compile some fortran code when it is installed with pip install numpy.

File: pythran/pythonic/numpy/dot.hpp (hunk 3 above)
Comment (pbrunet) [inline pythran/pythonic/numpy/dot.hpp]: I am just commenting the  `E::value == 1 and F::value == 1`

File: pythran/pythonic/numpy/dot.hpp (hunk 4 above)
Comment (pbrunet) [inline pythran/pythonic/numpy/dot.hpp]: because list + list is not a point-to-point operator.

File: pythran/pythonic/numpy/dot.hpp (hunk 5 above)
Comment (pbrunet) [inline pythran/pythonic/numpy/dot.hpp]: Yes it is.

File: pythran/pythonic/numpy/dot.hpp (hunk 6 above)
Comment (pbrunet) [inline pythran/pythonic/numpy/dot.hpp]: As I had some issue with gexpr (uncorrect results), I prefer keep it that way until you PR is merged.

File: pythran/toolchain.py (hunk 7 above)
Comment (pbrunet) [inline pythran/toolchain.py]: I hope so but I don't know as my numpy package is not compiled with openblas.

File: .travis.yml (hunk 2 above)
Comment (serge-sans-paille) [inline .travis.yml]: Yes, it does. Thanks!

File: pythran/pythonic/numpy/dot.hpp (hunk 3 above)
Comment (serge-sans-paille) [inline pythran/pythonic/numpy/dot.hpp]: using an horrible english I tried to correct ;-)

**Comment by serge-sans-paille** [comment]
@jfalcou : do you plan to do a release of nt2 soon?

**Comment by pbrunet** [comment]
@jfalcou , @serge-sans-paille This PR doesn't validate because nt2 now require boost 1.57 which is not available in debian testing/ubuntu for now (only 1.55 is available).
Should we monkey patch nt2 to only use the new  dgemv part? Any other ideas?

**Comment by serge-sans-paille** [comment]
@pbrunet : maybe @jfalcou could backport the relevant branch?

**Comment by serge-sans-paille** [comment]
@pbrunet this one should be closed, as it's superseded by #451 ?

**Pull request closed without merging**